### ISSUE TITLE
`destructuring-assignment` shouldn't trigger on method calls

### ISSUE THREAD

**Issue by rohit-gohri** [yannickcr/eslint-plugin-react#2960]
The rule `react/destructuring-assignment` is also triggered when using methods of params.

```jsx
      formatter: (val) => {
        return <span>{val.toPrecision(2)}</span>;
      },
```

<img width="706" alt="image" src="https://user-images.githubusercontent.com/31949290/113500966-b3fb4800-953f-11eb-970f-48328ac251f1.png">

Might also be related to #1950

**Comment by ljharb** [comment]
What is the context here? why is this an arrow function as an object key, and not a normal/concise function? How is this a react component if it has object keys?

**Comment by rohit-gohri** [comment]
Sorry for the incomplete description, Full component:

```jsx
import React from 'react';
import { OverviewCard, Tooltip } from '../components';

const Benchmark = ({
  benchmark,
}) => {
  const cardValues = [
    {
      key: 'kpi',
	  value: benchmark,
      formatter: (val) => {
        return (
          <Tooltip title="Compared to last month">
            <span>{val.toPrecision(2)}</span>
          </Tooltip>
        );
      },
    },
  ];

  return (
    <OverviewCard
      title="Title"
      values={cardValues}
    />
  );
};

export default Benchmark;
```

**Comment by ljharb** [comment]
While that should really be `formatter(val) {` instead of `formatter: (val) => {`, i agree that it shouldn’t be detected as a component.

**Comment by rohit-gohri** [comment]
Even with `formatter(val) {` it's still detecting it as a component.

<img width="817" alt="image" src="https://user-images.githubusercontent.com/31949290/113512370-fcd3f080-9581-11eb-8c9a-16e2ec402f49.png">

**Comment by stefan-toubia** [comment]
Also seeing this after upgrading to 7.23.2, it's not broken in 7.20.6

![image](https://user-images.githubusercontent.com/4265922/116793639-bdaf9700-aa7c-11eb-979d-c2e1b56eefbb.png)

Similar context as above

```js
const columns = [
  col1,
  {
    ...genVals("rule", "rule1"),
    render(rules) {
      const message = "a message";
      return (
        <Tooltip
          title={
            <div>
              <span>{message}</span>
              {rules.map((rule) => (
                <span className="rule" key={rule.id}>
                  {rule.text}
                </span>
              ))}
            </div>
          }
        >
        </Tooltip>
      );
    },
  },
  col3,
];
```

Doesn't matter if it's `render: (rules) => {` or `render(rules){`

**Comment by ljharb** [comment]
@stefan-toubia that's a component detection bug; a render function shouldn't have that triggered at all

**Comment by Wesitos** [comment]
Looks like it was introduced in commit 734dc5384c72a1ae2d5d4ddf0fffd2bb8d2abfd1

**Comment by Wesitos** [comment]
Found 2 false positives. But I'm not sure if they are related to this issue.
```js
const obj = {
  foo(val) {
    const a = val.func();
    return null;
  },
};
```
```js
export default (arg) => {
  const a = arg.foo();
  return null;
}
```

**Comment by ljharb** [comment]
@Wesitos the first one can't be a component, because it has a lowercased name; the second could be, and so it'd be reasonable to assume it needs propTypes.

**Comment by Wesitos** [comment]
@ljharb In the first case , looks like any method that returns null is detected as an `StatelessComponent`. I suppose the name of the function is not being checked for methods of object literals so it only depends on the return value of `isReturningJSXOrNull`.

Wouldn't it be reasonable to ignore functions that **always** return null?.

**Comment by ljharb** [comment]
No, since those could still be performing side effects.

**Comment by Wesitos** [comment]
I found a false negative related to component detection

```js
const Foo = ({prop}) => {
  if (prop) {
    return (
        <a>Something</a>
    );
  }

  return 'Another thing';
};
```
In this case `isReturningJSX` does not take into account the contents of `IfStatements` (Not sure if that is intended).

**Comment by ljharb** [comment]
It's not, that is indeed a false negative.

Component detection is always a WIP and could always use improvements; PRs are welcome.

**Comment by Wesitos** [comment]
> No, since those could still be performing side effects.

Sounds like a hook :stuck_out_tongue: 

> Component detection is always a WIP and could always use improvements; PRs are welcome.

I'll give it a check. Thanks

**Comment by Wesitos** [comment]
> @Wesitos the first one can't be a component, because it has a lowercased name; the second could be, and so it'd be reasonable to assume it needs propTypes/is a component/needs destructured props.

@ljharb Modifying `Components.getStatelessComponent` I noticed that this method was handling this old case (extracted from `destructuring-assignment` tests:

```js
var Hello = React.createClass({
   render: function() {
     return <Text>{this.props.foo}</Text>;
  }
})
```
Should that still be **always** detected as a component or only when modifying the `createClass` setting?

**Comment by ljharb** [comment]
@Wesitos `React.createClass` (assuming the pragma and createClass settings point at that) should always be detected as a component, yes.
